### PR TITLE
[Kernel][Metrics][PR#1] Adds initial interfaces and a Timer class

### PR DESCRIPTION
**Pull request by allisonport-db**
#### Which Delta project/connector is this regarding?

- [ ] Spark
- [ ] Standalone
- [ ] Flink
- [X] Kernel
- [ ] Other (fill in here)

## Description

Adds the initial interfaces for https://github.com/delta-io/delta/issues/3905 as well as a `Timer` class that will be used in follow-up PRs.

## How was this patch tested?

Just interface changes.

## Does this PR introduce _any_ user-facing changes?

No.

### CODE REVIEW

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/metrics/Timer.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    return new DefaultTimed(this);
+  }
+
+  /**
+   * Records a custom amount.
+   *
+   * @param amount The amount to record in nanoseconds
+   */
+  public void record(long amount) {
+    checkArgument(amount >= 0, "Cannot record %s %s: must be >= 0", amount);
```
Comment (scottsand-db) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/metrics/Timer.java:56]: Two `%s` but only one argument?

File: kernel/kernel-api/src/main/java/io/delta/kernel/metrics/DeltaOperationReport.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.delta.kernel.metrics;
+
+import java.util.Optional;
+import java.util.UUID;
+
+/** Defines the common fields that are shared by reports for Delta operations */
+public interface DeltaOperationReport extends MetricsReport {
```
Comment (scottsand-db) [inline kernel/kernel-api/src/main/java/io/delta/kernel/metrics/DeltaOperationReport.java:22]: Will `MetricsReporter::report` ever be called with a different type of `MetricsReport` that is _not_ a `DeltaOperationReport`?

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/metrics/Timer.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+package io.delta.kernel.internal.metrics;
+
+import static io.delta.kernel.internal.util.Preconditions.checkArgument;
+
+import java.util.concurrent.Callable;
+import java.util.concurrent.atomic.LongAdder;
+import java.util.function.Supplier;
+
+/** A timer class for measuring the duration of operations in nanoseconds */
+public class Timer {
```
Comment (scottsand-db) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/metrics/Timer.java:26]: Tests?

**Review by scottsand-db — COMMENTED** [review]
Looks great! Asked some questions.

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/metrics/Timer.java (hunk 1 above)
Comment (allisonport-db) [reply kernel/kernel-api/src/main/java/io/delta/kernel/internal/metrics/Timer.java:56]: Thanks good catch

File: kernel/kernel-api/src/main/java/io/delta/kernel/metrics/DeltaOperationReport.java (hunk 2 above)
Comment (allisonport-db) [reply kernel/kernel-api/src/main/java/io/delta/kernel/metrics/DeltaOperationReport.java:22]: Yes in the future possibly. We want `MetricsReport` to be completely generic so we can push anything we want as part of the interface.

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/metrics/Timer.java (hunk 3 above)
Comment (allisonport-db) [reply kernel/kernel-api/src/main/java/io/delta/kernel/internal/metrics/Timer.java:26]: Added a test suite

File: kernel/kernel-api/src/test/scala/io/delta/kernel/metrics/MetricsUtilsSuite.scala [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+import org.scalatest.funsuite.AnyFunSuite
+
+class MetricsUtilsSuite extends AnyFunSuite {
+
+  def millisToNanos(millis: Long): Long = {
+    millis*1000000
+  }
+
+  /**
+   * @param incrementFx Function given (duration, timer) increments the timer by about duration ms
```
Comment (scottsand-db) [inline kernel/kernel-api/src/test/scala/io/delta/kernel/metrics/MetricsUtilsSuite.scala:31]: nit: (non-blocking) "by approximately duration ms". "by about" sounds weird

**Review by scottsand-db — APPROVED** [review]
LGTM!

**Pull request merged**